### PR TITLE
[GSOC] Implement AudioOutput API

### PR DESCRIPTION
**Pull request by ETeissonniere**
As part of my GSOC proposal I have implemented a way to play audio files *from memory* on a `mettle` implant.

## Verification

- [ ] Get a session with a `mettle` implant connected
- [ ] Issue `play -f <path to test.wav>`
- [ ] If the implant can execute `aplay` you should hear your file playing.

### CODE REVIEW

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
+class Mic
```
Comment (mkienow-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb]: This looks like it was copied and pasted from `lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/mic.rb`, does this have any relation to this new module?

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+          Klass = Console::CommandDispatcher::Stdapi::AudioOutput
+
+          include Console::CommandDispatcher
+
+          #
+          # List of supported commands.
+          #
+          def commands
+            all = {
+              'play' => 'play an audio file on target\'s system, nothing written on disk'
```
Comment (mkienow-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb]: I'd suggest using double quotes for the description portion so you could drop the escape. 
Per [The Ruby Style Guide - Strings](https://github.com/bbatsov/ruby-style-guide#strings):
>Prefer double-quotes unless your string literal contains " or escape characters you want to suppress.

Comment (bcoles) [reply lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb]: `target system` also works

Comment (mkienow-r7) [reply lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb]: Rewording is fine as well. I figured I'd use this as an opportunity to introduce the style guide. ;)

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb (hunk 1 above)
Comment (DeveloppSoft) [reply lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb]: Whoops, fixing.

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb (hunk 2 above)
Comment (DeveloppSoft) [reply lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb]: Indeed! I have taken the rewording way, will keep in mind the style guide for the upcoming coding sessions.

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                print_line("Usage: audio_play [options]\n")
+                print_line("Upload file to targets memory and play it from memory")
+                print_line(play_start_opts.usage)
+                return
+              when "-f"
+                audio_path = val
+              end
+            end
+
+            print_status("Playing #{audio_path}...")
```
Comment (timwr) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb]: Minor: We should probably print the usage if audio_path is nil (e.g -f is omitted).

Comment (DeveloppSoft) [reply lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb]: Sure, fixing.

File: lib/rex/post/meterpreter/extensions/stdapi/audio_output/audio_output.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+require 'rex/post/meterpreter/channel'
+require 'rex/post/meterpreter/channels/pools/stream_pool'
+
+module Rex
+  module Post
+    module Meterpreter
+      module Extensions
+        module Stdapi
+          module AudioOutput
```
Comment (mkienow-r7) [inline lib/rex/post/meterpreter/extensions/stdapi/audio_output/audio_output.rb]: Minor: remove indention to be more consistent with other stdapi classes. For example, see `lib/rex/post/meterpreter/extensions/stdapi/fs/dir.rb`.

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb [hunk 5]
```diff
@@ -0,0 +1,74 @@
+# -*- coding: binary -*-
+require 'rex/post/meterpreter'
+require 'bindata'
+
+module Rex
+  module Post
+    module Meterpreter
+      module Ui
```
Comment (mkienow-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb]: Minor: remove indention to be more consistent with other stdapi classes. For example, see `lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/fs.rb`.

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        audio_path = val
+      end
+    end
+
+    if audio_path == nil
+      print_line('Please specify a path to an audio file via the -f option or use -h')
+      return
+    end
+
+    print_status('Playing #{audio_path}...')
```
Comment (mkienow-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb]: This requires double quotes for the expression substitution. It now displays `Playing #{audio_path}...`.

Comment (DeveloppSoft) [reply lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/audio_output.rb]: Whoops, fixing.

**Comment by timwr** [comment]
Any objections to landing this as-is and addressing any issues individually? We can always revert

**Comment by mkienow-r7** [comment]
@timwr That's fine. It can be iteratively improved and enhanced.

**Comment by timwr** [comment]
Ooops. I accidentally rebased your changes before landing them, so this pr wasn't closed automatically.
Closing manually now.
Many thanks @DeveloppSoft !!

**Comment by asoto-r7** [comment]
## Release Notes

mettle now includes a 'play' command which takes a -f parameter, pointing to a WAV file.  If 'aplay' is installed on the target, the WAV will be copied to the memory of the target and played via the default audio output.

**Pull request closed without merging**